onnxruntime/core/providers/coreml/coreml_execution_provider.cc

```diff
@@ -0,0 +1,314 @@
+          erased.insert(input);
+        }
+        //only when input is neither in output list nor erased list, add the input to input list
+        else if (erased.find(input) == erased.end()) {
+          fused_inputs[input] = input_order++;
+        }
+      }
+
+      // For output searching, there is a special case:
+      // If certain output is used more than once,
+      // if the output is connected to nodes that don't belong to the subgraph, the output need to be added
+      // to the output list
+
+      std::unordered_set<const NodeArg*> processed_outputs;
+      for (auto it = node->OutputEdgesBegin(), end = node->OutputEdgesEnd(); it != end; ++it) {
+        const auto node_idx = it->GetNode().Index();
+        const auto* output = node->OutputDefs()[it->GetSrcArgIndex()];
+
+        if (node_set.find(node_idx) != node_set.end()) {
+          const auto iter = fused_inputs.find(output);
+          if (iter != fused_inputs.end()) {
+            fused_inputs.erase(iter);
+            erased.insert(output);
+          } else if (erased.find(output) == erased.end()) {
+            fused_outputs[output] = output_order++;
+          }
+        } else {
+          fused_outputs_to_add[output] = output_order++;
+        }
+
+        processed_outputs.insert(output);
+      }
+
+      for (const auto* output : node->OutputDefs()) {
+        if (processed_outputs.find(output) != processed_outputs.end())
+          continue;
+
+        const auto iter = fused_inputs.find(output);
+        if (iter != fused_inputs.end()) {
+          fused_inputs.erase(iter);
+          erased.insert(output);
+        }
+        // only when output is neither in input list nor erased list, add the output to output list
+        else if (erased.find(output) == erased.end() && output->Exists()) {
+          fused_outputs[output] = output_order++;
+        }
+      }
+    }
+
+    fused_outputs.insert(fused_outputs_to_add.begin(), fused_outputs_to_add.end());
+    // Sort inputs and outputs by the order they were added
+    std::multimap<int, const NodeArg*> inputs, outputs;
+
+    for (auto it = fused_inputs.begin(), end = fused_inputs.end(); it != end; ++it) {
+      inputs.insert(std::pair<int, const NodeArg*>(it->second, it->first));
+    }
+
+    for (auto it = fused_outputs.begin(), end = fused_outputs.end(); it != end; ++it) {
+      if (all_node_inputs.find(it->first->Name()) != all_node_inputs.end()) {
+        outputs.insert(std::pair<int, const NodeArg*>(it->second, it->first));
+      } else if (std::find(graph_outputs.begin(), graph_outputs.end(), it->first) != graph_outputs.end()) {
+        outputs.insert(std::pair<int, const NodeArg*>(it->second, it->first));
+      }
+    }
+
+    // Assign inputs and outputs to subgraph's meta_def
+    auto meta_def = onnxruntime::make_unique<::onnxruntime::IndexedSubGraph::MetaDef>();
+    meta_def->name = "COREML_" + std::to_string(metadef_id_++);
+    meta_def->domain = kMSDomain;
+
```

Comment:
Do we need to use kMSDomain or is an arbitrary string fine? Given it's a compiled node we're never going to export it and we don't need to do kernel lookup for it.

Reply:
I don't think this matters, any string should work fine

---
In reply to: [563459806](https://github.com/microsoft/onnxruntime/pull/6392#discussion_r563459806) [](ancestors = 563459806)

Reply:
It's mildly confusing to use a semi-official domain where all our contrib ops are for arbitrary EPs that aren't necessarily created by/owned by MS (as this approach has been cut-and-pasted into multiple EPs). Should probably clean this up in all EPs at some point.

Reply:
Will create a separated EP domain for this in a later PR, such as "kFuntionBasedExecutionProvider" or "kCustomExecutionProvider"

---
In reply to: [565062217](https://github.com/microsoft/onnxruntime/pull/6392#discussion_r565062217) [](ancestors = 565062217)